mlflow/utils/autologging_utils.py

```diff
@@ -275,3 +282,202 @@ def batch_metrics_logger(run_id):
+                                    patch.
+    :param destination: The Python class on which the patch function is being defined.
+    :param function_name: The name of the function to patch on the specified `destination` class.
+    :param function: The patch function to apply. The first argument to this function should be
+                     reserved for an `original` method argument representing the underlying /
+                     original function. Subsequent arguments should be identical to those of the
+                     original function being patched.
+    """
+
+    def patched_train(*args, **kwargs):
+        preexisting_run = mlflow.active_run()
+        original_result = None
+        called_original = False
+        failed_during_original = False
+
+        def call_original(*og_args, **og_kwargs):
+            original = gorilla.get_original_attribute(destination, function_name)
+
+            if _is_testing():
+                _validate_args(args, kwargs, og_args, og_kwargs)
+
+            def wrapped_original(*args, **kwargs):
+                try:
+                    called_original = True
+                    nonlocal original_result
+                    original_result = original(*args, **kwargs)
+                    return original_result
+                except Exception:
+                    nonlocal failed_during_original
+                    failed_during_original = True
+                    raise
+
+            return wrapped_original(*og_args, **og_kwargs)
+
+        original = gorilla.get_original_attribute(destination, function_name)
+        call_original = functools.wraps(original)(call_original)
+        call_original.__signature__ = inspect.signature(original)
+
+        config = AUTOLOGGING_INTEGRATIONS[autologging_integration]
+        if config.get("disable", False):
+            return original(*args, **kwargs)
+
+        try:
+            return function(call_original, *args, **kwargs)
+        except Exception as e:
+            if _is_testing():
+                raise
+
+            if not preexisting_run and mlflow.active_run():
+                try_mlflow_log(mlflow.end_run, RunStatus.to_string(RunStatus.FAILED))
+
+            if failed_during_original:
+                raise
+
+            _logger.warning(
+                "Encountered unexpected error during %s autologging: %s", autologging_integration, e
+            )
+
+            if called_original:
+                return original_result
+            else:
+                return original(*args, **kwargs)
+
+    wrap_patch(destination, function_name, patched_train)
+
+
+_ATTRIBUTE_EXCEPTION_SAFE = "exception_safe"
+
+
+def exception_safe_function(function):
+    """
+    Wraps the specified function with broad exception handling to guard
+    against unexpected errors during autologging.
+    """
+    if _is_testing():
+        setattr(function, _ATTRIBUTE_EXCEPTION_SAFE, True)
```

Comment:
Leverages function attributes from https://www.python.org/dev/peps/pep-0232/ for argument validation purposes